Documentation: Add ServerContext, ServerConfiguration and GettingStartedWithDPFServer sections (#637)
pyansys/pydpf-core: docs/source/user_guide/server_types.rst

.. _user_guide_server_types:

===========================
Client-server communication
===========================

Terminology
-----------

> DPF is based on a **client-server** architecture. It allows **local** or **remote**

Reviewer: In the getting started section, I went for communication "in process" or "through the network" to be closer to the truth. We can use either one or the other but should be consistent maybe

Author: @PProfizi I took that comment into account, and the discussion we had yesterday; we can discuss it if needed.

.. _user_guide_server_types:

===========================
Client-server communication
===========================

Terminology
-----------

DPF is based on a **client-server** architecture. It allows **local** or **remote**
client-server **communication** capabilities.

The DPF Server is a set of files that enables DPF capabilities.

The different DPF client APIs (CPython, IronPython, C++, and so on) enable the use
of these capabilities.

PyDPF is a term encompassing both the CPython client
(the Python packages such as ansys-dpf-core or ansys-dpf-post, available on PyPI)
and the IronPython client (available within Ansys Mechanical).

> Getting started with DPF local server

Reviewer 2: Here we say local but next only gRPC, I think those titles need rewording
```suggestion
Starting DPF as a local server
```

> Getting started with DPF local server
> -------------------------------------

Reviewer 2: ```suggestion
Getting started with DPF as a local server
------------------------------------------
```

 .. code-block:: none
 
    <ansys.dpf.core.server_types.InProcessServer object at ...>

This server can now be used to instantiate Models, Operators, and so on.

.. code-block::
	
    # instantiate an operator
    local_operator = dpf.operators.results.displacement(server=local_server)
	
    # instantiate a model
    from ansys.dpf.core import examples
    local_model = dpf.Model(examples.find_simple_bar(), server=local_server)
	

Getting started with DPF GRPC server
------------------------------------

GRPC communication is enabled using :class:`GrpcServer <ansys.dpf.core.server_types>`. 

.. code-block::
	   
    from ansys.dpf import core as dpf
    grpc_server_config = dpf.AvailableServerConfigs.GrpcServer
    grpc_server = dpf.start_local_server(config=grpc_server_config)
    grpc_server

.. rst-class:: sphx-glr-script-out

 .. code-block:: none
 
    <ansys.dpf.core.server_types.GrpcServer object at ...>

You can obtain the server port and ip:

.. code-block::

    print(grpc_server)
	
.. rst-class:: sphx-glr-script-out

 .. code-block:: none

    DPF Server: {'server_ip': '127.0.0.1', 'server_port': 50052, 'server_process_id': 9999, 'server_version': '6.0', 'os': 'nt'}
	

> From a another machine, you can connect remotly to this server and instantiate Models, Operators, and so on:

Reviewer 2: ```suggestion
From another machine, you can connect remotely to this server and instantiate Models, Operators, and so on:
```

---------------------------------------

The different DPF server types can be started using one of the 
:class:`AvailableServerConfigs <ansys.dpf.core.server_factory>` configurations. 

.. code-block::
    
    in_process_config = dpf.AvailableServerConfigs.InProcessServer
    in_process_server = dpf.start_local_server(config=in_process_config)
    
    grpc_config = dpf.AvailableServerConfigs.GrpcServer
    grpc_server = dpf.start_local_server(config=grpc_config)
    
    legacy_grpc_config = dpf.AvailableServerConfigs.LegacyGrpcServer
    legacy_grpc_server = dpf.start_local_server(config=legacy_grpc_config)

Advanced concepts and release history
-------------------------------------

The communication logic with a DPF server is defined when starting it using
an instance of the :class:`ServerConfig <ansys.dpf.core.server_factory>` class.
Different predefined server configurations are available in DPF,
each answering a different use-case
(See the :class:`AvailableServerConfigs <ansys.dpf.core.server_factory>` class).

- The :class:`GrpcServer <ansys.dpf.core.server_types>` configuration is available starting 
  with server version 4.0 (Ansys 2022 R2).

>   It allows you to remotely connect to a DPF server across a network by telling the client

Reviewer 2: Now that I'm reading it again, saying "you" seems strange.
```suggestion
  It allows to remotely connect to a DPF server across a network by telling the client
```

Reviewer 3: No, "it allows you" is correct. Otherwise the sentence is missing a piece. It allows (what?) to remotely connect...

  to communicate with this server via the gRPC communication protocol.

>   Although it can be used to communicate with a DPF server running on the same local machine,
>   in that case the next configuration is better for this option.

Reviewer 3: ```suggestion
  Although it can be used to communicate with a DPF server running on the same local machine, the next configuration is better for this option.
```
